### PR TITLE
feat(constraint_graph): make the constraint graph framework generic and move it into a separate crate

### PR DESCRIPTION
**Pull request by vspecky**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR separates out and refactors the Euclid Knowledge Graph into the Constraint Graph in its own crate, and makes the Constraint Graph generic so it can work on custom key and value types as opposed to just the Euclid Key and Value types.

This will remove the dependency on Euclid when one needs to only use the Constraint Graph, plus the now generic nature of Constraint Graphs will allow them to be utilized for varied use cases other than routing.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Making the constraint graph generic opens up the opportunity for other developers to adapt and use the constraint graph for their own usecases. There are many validation use cases currently in Hyperswitch that can be fulfilled by using Constraint Graphs, including but not limited to Connector domain validations, required field selection, etc.

## How did you test it?

Unit tests for the `euclid` and `kgraph_utils` crates. No specific API tests required.

Euclid:
image

Kgraph_utils:
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by prajjwalkumar17 — DISMISSED** [review]
LGTM

**Review by Aprabhat19 — DISMISSED** [review]

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by jarnura — DISMISSED** [review]

**Review by jarnura — APPROVED** [review]

**Review by Aprabhat19 — APPROVED** [review]

File: crates/hyperswitch_constraint_graph/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "hyperswitch_constraint_graph"
+description = "Constraint Graph Framework for modeling Domain-Specific Constraints"
+version = "0.1.0"
+edition.workspace = true
+rust-version.workspace = true
```
Comment (SanchithHegde) [inline crates/hyperswitch_constraint_graph/Cargo.toml:6]: Can we add a `license` and optionally, a `readme` field?

**Review by SanchithHegde — APPROVED** [review]
Other than that, looks good to me!

File: crates/hyperswitch_constraint_graph/Cargo.toml (hunk 1 above)
Comment (pixincreate) [reply crates/hyperswitch_constraint_graph/Cargo.toml:6]: Will add the PR to the merge queue as @vspecky confirmed to add the license and readme in the next upcoming PR.

**Pull request merged**